### ISSUE TITLE
Take DataView concept into use

### ISSUE THREAD

**Issue by caalador** [vaadin/vaadin-combo-box-flow#355]
Based on vaadin/flow#8303 and the common implementation parts introduced there, take ListDataView into use with ComboBox. Make a PR to feature/dataprovider branch, use the 3.2.dataprovider-SNAPSHOT version of Flow.

The demo for data view should showcase:

 - migrating all existing demos to use data view-based API instead
 - setting items and creating next/previous controls
 - filtering and sorting

**Comment by mstahv** [comment]
Removed sorting from the list as there is no sorting in ComboBox (but what developer does).

**Comment by mshabarov** [comment]
As it figured out during the initial attempt to implement `HasXXXDataView` to `ComboBox`, there are two obstacles to do that smoothly:

- `HasDataView` contains `setItems(InMemoryDataProvider<T> dataProvider)` which does not match with an existing `setDataProvider` methods, because `InMemoryDataProvider` has a `SerializablePredicate<T>` as a filter type and `ComboBox` has a `String`
- `HasLazyDataView` has a `Void` as a filter type, so it cannot be applied to `ComboBox` as is.

The first issue was resolved just by skipping the `HasDataView` for `ComboBox`, and the second required to create a new interface `HasFilterableLazyDataView` with an additional generic type representing the filter type:

```
public interface HasFilterableLazyDataView<T, F, V extends FilterableLazyDataView<T, F>>
        extends Serializable {

    default V setItemsWithFilter(
            CallbackDataProvider.FetchCallback<T, F> fetchCallback) {
        return setItemsWithConvertedFilter(fetchCallback,
                SerializableFunction.identity());
    }

    default V setItemsWithFilter(
            CallbackDataProvider.FetchCallback<T, F> fetchCallback,
            CallbackDataProvider.CountCallback<T, F> countCallback) {
        return setItemsWithFilter(DataProvider
                .fromFilteringCallbacks(fetchCallback, countCallback));
    }

    default <Q> V setItemsWithConvertedFilter(
            CallbackDataProvider.FetchCallback<T, Q> fetchCallback,
            SerializableFunction<F, Q> filterConverter) {
        V filterableLazyDataView = setItemsWithConvertedFilter(fetchCallback,
                query -> {
                    throw new IllegalStateException(
                            "Trying to use exact size with a lazy loading component"
                                    + " without either providing a count callback for the"
                                    + " component to fetch the count of the items or a data"
                                    + " provider that implements the size query. Provide the "
                                    + "callback for fetching item count with%n"
                                    + "component.getFilterableLazyDataView().withDefinedSize(CallbackDataProvider.CountCallback);"
                                    + "%nor switch to undefined size with%n"
                                    + "component.getFilterableLazyDataView().withUndefinedSize();");
                }, filterConverter);
        filterableLazyDataView.setItemCountUnknown();
        return filterableLazyDataView;
    }

    V setItemsWithFilter(BackEndDataProvider<T, F> dataProvider);

    <Q> V setItemsWithConvertedFilter(
            CallbackDataProvider.FetchCallback<T, Q> fetchCallback,
            CallbackDataProvider.CountCallback<T, Q> countCallback,
            SerializableFunction<F, Q> filterConverter);

    V getFilterableLazyDataView();
}
```

This new interface provides the similar `setItems` overloads as a `HasLazyDataView`, but with additional filter converter callback.
`FilterableLazyDataView` has been added either, to provide `setFilter`/`removeFilter` API:

```
public interface FilterableLazyDataView<T, F> extends LazyDataView<T> {

    void setFilter(F filter);

    default void removeFilter() {
        setFilter(null);
    }

    void setFilterCombiner(FilterCombiner<F> filterCombiner);

    void setItemCountCallbackWithFilter(
            CallbackDataProvider.CountCallback<T, F> callback);
}
```

It's implementation implied to use a `filterSlot` to set the filter programmatically:

```
SerializableConsumer<String> filterSlot = filter -> {
                if (!Objects.equals(filter, lastFilter)) {
                    providerFilterSlot.accept(convertOrNull.apply(filter));
                    lastFilter = filter;
                }
            };
```

This approach was a solution for the first glance, but when I started to create a IT tests, I experienced a one more implementation detail of `ComboBox` - the client side filter is set each time the user opens the popup. That means the programmatic filter does not make sense because the client filter will replace it each time anyway. I also considered an option to separate the client and configured filter from each other and to create a new field for configured filter, but, anyway, these two filters have to be merged before being put to the Query object and, thus, it's not much valuable.

So now it looks like the `setFilter` isn't something really useful for ComboBox. However, it would be useful for other filtered lazy loading components.

Saying that, I have a doubts about how to proceed with a `ComboBox` and filtering API:

- To keep the `setFilter/removeFilter` in the API, but throw an UnsupportedOperationException for ComboBox? This would be a bit weird, because ComboBox is a main candidate for this API and it doesn't support it. Super weird DX.
- To remove the `setFilter/removeFilter`? But how about `Grid` and `IronList`? They potentially can also implement filtering API.
- To change the `ComboBox` drastically so as to support the programmatic filtering.

IMO it requires a design review and DX tests.

**Comment by mshabarov** [comment]
After discussing how the `ComboBox` could handle the programmatic filtering API, we decided to keep `setFilter/removeFilter` methods and add a filter combiner so as to merge the text filter with a custom filter and to convert it to query filter. These are the steps I made to achieve that:
1. `FilterableLazyDataView` contains API for programmatic filter handling:

```
public interface FilterableLazyDataView<T, F> extends LazyDataView<T> {

    void setFilter(Object filter);

    default void removeFilter() {
        setFilter(null);
    }

    void setItemCountCallbackWithFilter(
            CallbackDataProvider.CountCallback<T, F> callback);
}
```

2. `HasFilterableLazyDataView` contains two methods for setting the callbacks with a custom filter type and filter combiner:

```
default <Q> V setItems(
            CallbackDataProvider.FetchCallback<T, Q> fetchCallback,
            FilterCombiner<F, Q> filterCombiner) {
// call to overloaded setItems
}

    <Q> V setItems(
            CallbackDataProvider.FetchCallback<T, Q> fetchCallback,
            CallbackDataProvider.CountCallback<T, Q> countCallback,
            FilterCombiner<F, Q> filterCombiner);
```

3. `FilterCombiner` is a `BiFunction` with the following types:

```
@FunctionalInterface
public interface FilterCombiner<F, Q> extends SerializableBiFunction<F, Object,
        Q> {

    static <F> FilterCombiner<F, F> identity() {
        return (componentFilter, customFilter) -> componentFilter;
    }
}
```

It combines component filter `F` (`String` in case of `ComboBox`) and the custom filter of type `Object`. Then it convert the object to custom type `Q`.

4. On a `ComboBox` side `setItems` method delegates to `setDataProvider` which is now modified as follows:

```
public <Q> void setDataProvider(DataProvider<T, Q> dataProvider,
            FilterCombiner<String, Q> filterCombiner)
```

Internally, `setDataProvider` method mixes the text filter and programmatic filter and pass the result to the `DataCommunicator`.

5. `ComboBox` is supplemented with `private Object customFilter` and `private FilterCombiner<String, ?> filterCombiner;` fields. `FilterCombiner` returns a text filter by default.

6. Here is an example how the filter method can be used:

```
ComboBoxFilterableLazyDataView<Person> dataView = comboBox
                .setItems(
                        query -> service.fetchWithCustomFilter(
                                // getFilter return a Person object
                                query.getFilter().orElse(null), 
                                query.getPage(),
                                query.getPageSize()),

                        query -> service.countWithCustomFilter(
                                query.getFilter().orElse(null)),

                        // Filter combiner: merge the text filter from client and the custom filter
                        (textFilter, customFilter) -> new Person(id, textFilter,
                                textFilter, ((Person) customFilter).getAge(),
                                anyAddress, anyPhoneNumber));

...
// Somewhere in change value listener of any other input component
// Set Person with an age as a filter
dataView.setFilter(new Person(0, null, null, selectedAge, null, null));
```

Further steps would be to review and conclude the `FilterableLazyDataView` methods so as to return a `DataView` with  `setFilter` only when the developer chooses `setItem` with `filterCombiner` overload.

**Comment by pleku** [comment]
> 2. HasFilterableLazyDataView contains two methods for setting the callbacks with a custom filter type and filter combiner:

Uhm, but not everyone needs the combiner so this is missing the basic methods which just take the callback where query filter type is `String` ?

> 4. On a ComboBox side setItems method delegates to setDataProvider which is now modified as follows:

We preferably should not modify the existing `setDataProvider` methods. I'm not sure why do we need to do this ?


I'm not sure if having the `setFilter` API with type `Object` is really that beneficial. It might help in the sense that the users could then call that when the (external) filter changes to trigger an update, but it is not really that much more useful compared to just calling `component.getDataView().refreshAll();` and then taking the external filter into account in the `FetchCallback`.
Having the filter API even makes it more complex to combine things with the case of `ComboBox`. Currently CB has API for "converting the internal filter" and then the data provider has the API for "combining the filter to another". After all I'm starting to think that we could both:
1) ditch the API for combining the component in-built filter (`String` in `ComboBox`, no other component currently) completely. Instead we keep the API with the filter converter
```java
setItems(FetchCallback<T, C> fetchCallback, SerializableFunction<String, C> filterConverter)
```
2) not add a `setFilter` API to the lazy data view, for now. We can add this later on if necessary
3) Being able to use `HasLazyDataView` in both `Grid` and `ComboBox`. In `Grid` the filter type parameter should be `Void` and in `ComboBox` it should be `String`. I was not able to achieve this without compiler warnings about unnecessary type parameter or by adding the filter type as a third type parameter for the interface:
```java
HasLazyDataView<T, V extends LazyDataView<T>, F>
```
With `F` being the filter type in the `setItems` methods. While it looks uglier and would require a new release for Flow & Grid for 17 beta, I think it still has the benefits of being able to reuse code better in the components, even though it will likely be just us.


----------

So the above would be my suggestion, please point where I'm off track @mshabarov @caalador.
Further requirements for this work should be:
- any existing (custom) `setItems` in CB are changed to return a data view of the correct type
- there is a replacement API with data view for all API which has `setDataProvider`
- something else I forgot ...

**Comment by mshabarov** [comment]
> Uhm, but not everyone needs the combiner so this is missing the basic methods which just take the callback where query filter type is String ?

Yes, sorry, I meant 'two methods among the others'. It should contain those basic methods for sure.

> We preferably should not modify the existing setDataProvider methods. I'm not sure why do we need to do this ?

Ok, we could keep

```
setDataProvider(DataProvider<T, C> dataProvider, SerializableFunction<String, C> filterConverter)
```

signature as it is, but then create a very similar implementation for `setItem` overload, 
but with filter combiner used (instead of fileConverter, which appropriate only for text filter).
BUT, we don't do that, of course, if we decide not to use custom filter and filter combiner.


> I'm not sure if having the setFilter API with type Object is really that beneficial.

I gave a try to make the filter type in `setFilter` generic, but as it looks now, we can't do that unless we have 
a custom filter generic type in the class declaration, which is not an appropriate apparently, because of heavy breaking change.

I generally agree with your proposal, just a couple of thoughts:

- So, we're going to use the existing `HasLazyDataView` and `LazyDataView` interfaces, but add a `F` to declaration, right?
- No more `FilterableLazyDataView` and `HasFilterableLazyDataView`
- `setItems(FetchCallback<T, C> fetchCallback, SerializableFunction<String, C> filterConverter)` is going to be included into `LazyDataView`. Does that mean the `filterConverter` will work as a supplier in `Grid`? Like this:

```
final Person personFilter = new Person("foo");
grid.setItems(query -> ..., ignoredVoid -> personFilter);
```
and an implementation would be similar to one in `ComboBox`, except there wouldn't be any filter slots.

**Comment by pleku** [comment]
> So, we're going to use the existing HasLazyDataView and LazyDataView interfaces, but add a F to declaration, right?

Agreed to do this only for `HasLazyDataView` and `LazyDataView` would only have it for `setItemCountCallback` which is fine to be `Object` for all components, thus changed to `?` from `Void`

> No more FilterableLazyDataView and HasFilterableLazyDataView

No. We can consider these later on if needed.

> setItems(FetchCallback<T, C> fetchCallback, SerializableFunction<String, C> filterConverter) is going to be included into LazyDataView

No. There would be no point in that since `Grid` does not have any inbuilt filter. But ComboBox should have this method as surplus to the ones from `HasLazyDataView` since it is useful for the component (and already exists with `DataProvider` (which is deprecated now)).

**Comment by mshabarov** [comment]
Acceptance criteria:
- `HasLazyDataView` has been updated with a new filter type, then update `void setItemCountCallback(
            CallbackDataProvider.CountCallback<T, Void> callback);` with wildcard as a filter parameter.
- Javadoc has been updated for `setItemCountCallback` so as refer the internal filter case in components and shift in to the end of class
- The tests on Flow side have been updated according to this newly added filter type
- `Grid` has been updated with a new `HasLazyDataView` and place the `Void` type there, update the Demo and tests
- ComboBox now does not implement `HasFilterableDataProvider`, but implements `HasLazyDataView` and has an additional `setItems` method with filterConverter:

```
setItems(FetchCallback<T, C> fetchCallback, SerializableFunction<String, C> filterConverter)
```

`setDataProvider` methods are now deprecated.
- ComboBox contains the lazy case unit and IT tests for filtering cases in define and undefined case.
- New Flow and Grid versions should be ported to `V17`, and ComboBox changes should go to `V18`.

**Comment by alvarezguille** [comment]
Solving #386 should be part of acceptance criteria for this new feature
